content/client-lib-development-guide/features.textile

```diff
@@ -465,10 +465,23 @@ h3(#realtime-channel). Channel
+* @(RTP2)@ A @PresenceMap@ should be used to maintain a list of members present on a channel. Broadly, this is is a map of "memberKeys":#TP3h to presence messages, all with @PRESENT@ actions (during a sync there may also be ones with an @ABSENT@ action, see "RTP2f":#RTP2f ). Consult the "the Java implementation":https://github.com/ably/ably-java/blob/master/lib/src/main/java/io/ably/lib/realtime/Presence.java#L578-L723 for details.
+** @(RTP2a)@ All incoming presence messages must be compared for newness with the matching member already in the @PresenceMap@, if one exists, where "matching" means they share the same @memberKey@ (or equivalently, they share both @connectionId@ and @clientId@)
+** @(RTP2b)@ To comparing for newness:
+** @(RTP2b1)@ Split the @id@ of both presence messages (which will be of the form @connid:msgSerial:index@, e.g. @aaaaaa:0:0@) on the separator @:@, to give @connectionId@, @msgSerial@, and @index@, parsing the latter two as integers (either eagerly or if/when needed)
+** @(RTP2b2)@ If the @connectionId@s are different, compare numerically by @timestamp@
+** @(RTP2b3)@ If the @connectionId@s are the same, compare first by @msgSerial@ numerically and then (if @msgSerial@s are equal) by @index@ numerically, larger being newer in both cases
+** @(RTP2c)@ As there are no guarantees that during a @SYNC@ operation presence events will arrive in order, all presence messages from a @SYNC@ must also be compared for newness in the same way as they would from a @PRESENCE@
+** @(RTP2d)@ When a presence message with an action of @ENTER@, @UPDATE@, or @PRESENT@ arrives, it should be added to the presence map with the action set to @PRESENT@
+** @(RTP2e)@ If a @SYNC@ is not in progress, then when a presence message with an action of @LEAVE@ arrives,  that @memberKey@ should be deleted from the presence map, if present
+** @(RTP2f)@ If a @SYNC@ is in progress, then when a presence message with an action of @LEAVE@ arrives, it should be stored in the presence map with the action set to @ABSENT@. When the @SYNC@ completes, any @ABSENT@ members should be deleted from the presence map. (This is because in a @SYNC@, we might receive a @LEAVE@ before the corresponding @ENTER@).
+** @(RTP2g)@ Any incoming presence message that results in a material change to the presence map (that is: which passes the newness check, and is not a @LEAVE@ message for a member which was not in the presence map or which was in the presence map with an @ABSENT@ action) should be emitted on the @Presence@ object, with an event name set to its original action. Note: this action may not be the same one that it will have when stored in the presence map. For example: an incoming presence message with an @ENTER@ action will be emitted as an @enter@ event, and the emitted presence message will have its action set to @ENTER@. However, it will be stored in the presence map with a @PRESENT@ action.
```

Comment:
Yep, we didn't discuss this - this one was just documenting the current behaviour (for ably-js anyway, I guess not ably-ruby :grinning:). That behaviour was added in [this commit](https://github.com/ably/ably-js/commit/9c4f8d78) to ably-js.

The idea (I'm guessing, @paddybyers can correct me if I'm wrong) is that we can encapsulate the oddness of leaves-arriving-before-presents (due to RTP2f behaviour) inside the library. I.e. someone subscribing to presence events will never see a LEAVE before the corresponding ENTER/PRESENT. (The LEAVE won't be shown because it's a leave for someone not in a present map; then the ENTER won't be shown because it'll fail the newness check against the ABSENT). So we can present a 'cleaned-up' presence interface which actually behaves like a set, in the sense that someone doing things using presence events emitted by the library will not have to worry about implementing tombstones (ABSENT actions); doing the 'naive thing' will just work.

WDYT?